boa/src/syntax/parser/mod.rs

```diff
@@ -1,916 +1,1733 @@
+    }
+
+    /// Checks if the lexical declaration continues with more bindings.
+    ///
+    /// If it does, it will advance the internal cursor to the next identifier token.
+    /// A Lexical Declaration continues its binding list if we find a `,` character. A New line
+    /// indicates the same as a `;`.
+    ///
+    /// More information: <https://tc39.es/ecma262/#prod-LexicalDeclaration>.
+    fn lexical_declaration_continuation(&mut self) -> Result<bool, ParseError> {
+        if let Some(tok) = self.cursor.peek(0) {
+            match tok.kind {
+                TokenKind::LineTerminator => {
+                    let _ = self.cursor.next().ok_or(ParseError::AbruptEnd)?;
+                    Ok(false)
+                }
+                TokenKind::Punctuator(Punctuator::Semicolon) => Ok(false),
+                TokenKind::Punctuator(Punctuator::Comma) => {
+                    let _ = self
+                        .next_skip_lineterminator()
+                        .ok_or(ParseError::AbruptEnd)?;
+                    Ok(true)
+                }
+                _ => Err(ParseError::Expected(
+                    vec![
+                        TokenKind::Punctuator(Punctuator::Semicolon),
+                        TokenKind::LineTerminator,
+                    ],
+                    self.cursor.next().ok_or(ParseError::AbruptEnd)?.clone(),
+                    Some("lexical declaration"),
+                )),
             }
-            TokenData::Punctuator(Punctuator::Arrow) => {
-                self.pos += 1;
-                let mut args = Vec::with_capacity(1);
-                match result.def {
-                    ExprDef::Local(ref name) => {
-                        args.push(Expr::new(ExprDef::Local((*name).clone())))
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// <https://tc39.es/ecma262/#prod-VariableDeclaration>
+    fn read_variable_declaration(&mut self) -> Result<(String, Option<Node>), ParseError> {
+        let tok = self
+            .next_skip_lineterminator()
+            .ok_or(ParseError::AbruptEnd)?;
+        let name = if let TokenKind::Identifier(name) = &tok.kind {
+            name.clone()
+        } else {
+            return Err(ParseError::Expected(
+                vec![TokenKind::Identifier("identifier".to_string())],
+                tok.clone(),
+                Some("variable declaration"),
+            ));
+        };
+
+        if self
+            .next_if(TokenKind::Punctuator(Punctuator::Assign))
+            .is_some()
+        {
+            Ok((name, Some(self.read_initializer()?)))
+        } else {
+            Ok((name, None))
+        }
+    }
+
+    /// Reads an initializer of variables.
+    ///
+    /// Note: it will expect the `=` token to have been already read.
+    ///
+    /// More information: <https://tc39.es/ecma262/#prod-Initializer>
+    fn read_initializer(&mut self) -> ParseResult {
+        self.read_assignment_expression()
+    }
```

Comment:
I added some information in the docs here, but it feels a bit contra-intuitive. The `Initializer` is always a `=` character followed by an `AssignmentExpression`. So I think that if we want to divide the logic, we shouldn't ask for the calling code to already parse part of the initializer (the `=` token, in this case).

I think we should change it to expect reading the `=` token, and then the assignment expression.

More info: https://tc39.es/ecma262/#prod-Initializer